include/cell.tcc

```diff
@@ -1236,10 +1236,58 @@ void mpm::Cell<Tdim>::compute_nodal_momentum(const Eigen::VectorXd& shapefn,
     nodes_[i]->update_momentum(true, phase, shapefn(i) * pmass * pvelocity);
 }
 
+//! Compute Bbar matrix
+template <unsigned Tdim>
+std::vector<Eigen::MatrixXd> mpm::Cell<Tdim>::compute_bbar(
+    const std::vector<Eigen::MatrixXd>& bmatrix, unsigned phase) {
+
+  // Compute volumetric Bmatrix
+  std::vector<Eigen::MatrixXd> bmatrix_vol;
+  for (unsigned i = 0; i < this->nnodes(); ++i) {
+    Eigen::Matrix<double, 3 * (Tdim - 1), Tdim> bvoli;
```

Comment:
Replace this with a computed final `bbar` matrix, that is just iterated over nodes and appended. This would make the calculation quicker.

Use template specialization to split between 2D and 3D version